### PR TITLE
Add check_remote_head option to avoid unnecessary new releases by che…

### PR DESCRIPTION
**Pull request by ahmadmayahi**
Checking the remote git HEAD without cloning the repo.

| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No
| Deprecations? | No
| Fixed tickets | #1755

A new option `check_remote_head` has been added, when this option sets to `true` deployer will not clone the repository if there are no changes found.

Thanks to `git ls-remote`, this command gets the `HEAD` without cloning the repository.

Revision is also checked, so if we run something like `dep deploy --revision=a69539703d917cafb1c51dad20f4ee1e317c3017` deployer will also check for the revision.

### CODE REVIEW

**Comment by antonmedv** [comment]
Build failed

**Comment by ahmadmayahi** [comment]
> Build failed

It keeps saying `Deployer\Support\Changelog\ParseException: Error in compare link syntax` I have no idea why?

**Comment by antonmedv** [comment]
Try checkout CHANGELOG and run `php bin/changelog`

**Comment by ahmadmayahi** [comment]
I tried that, but still get the same result:
Image of Yaktocat

**Comment by antonmedv** [comment]
Really strange, will look at this tonight.

**Comment by ahmadmayahi** [comment]
Thank you @cafferata

File: recipe/deploy/prepare.php [hunk 1]
```diff
@@ -36,4 +36,26 @@
...
+    if (true === get('check_remote_head') && null == input()->getOption('tag')) {
+        $headPath = trim(get('deploy_path').'/.dep/HEAD');
+        $isRemoteHeadExists = (bool) run('if [ -f '.$headPath.' ]; then echo 1; else echo 0; fi');
+
+        //check if HEAD file is exists and then compare it
+        if (true === $isRemoteHeadExists) {
+            $headContents = run('cat '.$headPath);
+            if ($headContents === $remoteHead) {
+                writeln('<info>Already up-to-date.</info>');
+                exit(0);
```
Comment (antonmedv) [inline recipe/deploy/prepare.php]: I think exiting program like this is dangerous and not flexible. What is some one want's to add hook what will be triggered all the time? Maybe throw GrasefulSutdown excetion?

File: recipe/deploy/prepare.php [hunk 2]
```diff
@@ -36,4 +36,26 @@
...
+    // Check and save the remote HEAD/revision and compare it with the existing saved one (if any)
+    // This avoid unnecessary releases when the last commit id matches the existing one (HEAD)
+    $repository  = trim(get('repository'));
+    $revision    = input()->getOption('revision') ?? null;
+    $remoteHead  = $revision ? $revision :
+        trim(run(sprintf('%s ls-remote %s HEAD | tr -d "HEAD"',
+            get('bin/git'), $repository)));
+    if (true === get('check_remote_head') && null == input()->getOption('tag')) {
+        $headPath = trim(get('deploy_path').'/.dep/HEAD');
+        $isRemoteHeadExists = (bool) run('if [ -f '.$headPath.' ]; then echo 1; else echo 0; fi');
```
Comment (antonmedv) [inline recipe/deploy/prepare.php]: You can use `test()` fn here. Also better to use double quotes `"[ -f $headPath.]"` as it more readable.

File: recipe/deploy/prepare.php [hunk 3]
```diff
@@ -36,4 +37,25 @@
...
+    // Check and save the remote HEAD/revision and compare it with the existing saved one (if any)
+    // This avoid unnecessary releases when the last commit id matches the existing one (HEAD)
+    $repository  = trim(get('repository'));
+    $revision    = input()->getOption('revision') ?? null;
+    $remoteHead  = $revision ? $revision :
+        trim(run(sprintf('%s ls-remote %s HEAD | tr -d "HEAD"',
+            get('bin/git'), $repository)));
+    if (true === get('check_remote_head') && null == input()->getOption('tag')) {
+        $headPath = trim(get('deploy_path').'/.dep/HEAD');
+        $isRemoteHeadExists = test("[ -f $headPath ]");
```
Comment (staabm) [inline recipe/deploy/prepare.php]: Is this check really required, or could it be achieved in combination with the „cat“ line in a single command?
Every remote command has a big overhead

Comment (ahmadmayahi) [reply recipe/deploy/prepare.php]: Well, `cat`ing non-existing file leads to `STDERR`.
I know, we could output the error to the `2>/dev/null` but I think it's a hack, the right solution is to check whether the file exists or not.

Comment (staabm) [reply recipe/deploy/prepare.php]: And something along the lines of
`run('test -f '. $headPath ' && cat '.$headPath);`

Comment (ahmadmayahi) [reply recipe/deploy/prepare.php]: This is great.

Comment (antonmedv) [reply recipe/deploy/prepare.php]: Nope, the https://en.wikipedia.org/wiki/Test_(Unix)

**Comment by staabm** [comment]
ping @antonmedv

**Pull request merged**